### DISCUSSION TITLE
How to use the Keras API (callbacks) in E2E case

### DISCUSSION THREAD

**Discussion by Mio-TW** [NVlabs/sionna#290] [Q&A] [answered]
Hi everyone,

I want to use the built-in functions of Keras for E2E model training (this [example](https://nvlabs.github.io/sionna/examples/Autoencoder.html#Trainable-End-to-end-System:-Conventional-Training))
For example:
tf.keras.callbacks.ReduceLROnPlateau
tf.keras.callbacks.EarlyStopping

But I don’t know how to add it. 
In addition, the validation case is not defined in the [example](https://nvlabs.github.io/sionna/examples/Autoencoder.html#Trainable-End-to-end-System:-Conventional-Training). 
How can I add it?

Thank you so much!

**Comment by jhoydis** [comment, marked as answer]
Hi, I think that your question is not specific to Sionna but rather to custom training loops in Keras.
For example, you can have a look at the [TF documentation](https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch) . There are also many blog posts that explain custom training loops with custom callbacks, e.g. [this one](https://alexander-pelkmann.medium.com/custom-training-with-custom-callbacks-3bcd117a8f7e).
